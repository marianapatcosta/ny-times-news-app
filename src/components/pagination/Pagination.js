import { defineComponent } from "vue";

export default defineComponent({
  name: "Pagination",
  emits: ["update-pagination"],
  props: {
    pageCount: {
      type: Number,
      required: true,
    },
  },
  data() {
    return {
      selectedPage: 0,
    };
  },
  computed: {
    paginationNumbers() {
      const nrButtonsToDisplay = window?.matchMedia("(min-width: 525px)")
        .matches
        ? 10
        : 5;
      const numbersOffset = 3;
      const numbers = Array(nrButtonsToDisplay).fill("");
      let firstNumberToDisplay = this.selectedPage;
      // determine the number option to display when selectedPage is near to pageCount (last page)
      if (this.selectedPage >= this.pageCount - numbersOffset) {
        firstNumberToDisplay = this.pageCount - nrButtonsToDisplay + 1;
      } else if (this.selectedPage >= nrButtonsToDisplay - numbersOffset) {
        firstNumberToDisplay =
          this.selectedPage - (nrButtonsToDisplay - numbersOffset);
      }

      return numbers.map((number, index) => firstNumberToDisplay + index);
    },
  },
  methods: {
    handlePaginationUpdate(newPage) {
      this.selectedPage = newPage;
      this.$emit("update-pagination", newPage);
    },
  },
});
