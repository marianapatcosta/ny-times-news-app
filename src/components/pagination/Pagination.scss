.pagination {
  background-color: $color-secondary;
  box-shadow: 0 0.125rem 0.5rem rgba($color-black, 0.7);
  display: inline-block;
  padding: 0.1rem 0.5rem;
  user-select: none;

  &__button {
    cursor: pointer;
    color: $color-primary;
    margin: 0 0.15rem;
    padding: 0.25rem;

    &--active {
      background-color: $color-primary;
      color: $color-secondary;
      font-weight: 700;
    }

    &:hover {
      transform: scale(1.25);
      -webkit-transform: scale(1.25);
      -moz-transform: scale(1.25);
      -ms-transform: scale(1.25);
      -o-transform: scale(1.25);
}

    &:disabled {
      cursor: revert;

      &:hover {
        transform: revert;
        -webkit-transform: revert;
        -moz-transform: revert;
        -ms-transform: revert;
        -o-transform: revert;
      }
    }
  }
}


