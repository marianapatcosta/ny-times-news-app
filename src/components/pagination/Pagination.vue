<template>
  <div class="pagination" data-cy="pagination">
    <button
      id="button-to-first-page"
      class="pagination__button"
      :aria-label="`Click to select first page`"
      :disabled="selectedPage === 0"
      @click="handlePaginationUpdate(0)"
      @keydown.enter="handlePaginationUpdate(0)"
    >
      &lt; &lt;
    </button>
    <button
      id="button-to-previous-page"
      class="pagination__button"
      :aria-label="`Click to select the previous page`"
      :disabled="selectedPage === 0"
      @click="handlePaginationUpdate(selectedPage - 1)"
      @keydown.enter="handlePaginationUpdate(selectedPage - 1)"
    >
      &lt;
    </button>
    <button
      v-for="number in paginationNumbers"
      class="pagination__button"
      :class="{
        'pagination__button--active': number === selectedPage,
      }"
      :key="`button-${number}`"
      :aria-label="`Click to select page ${number + 1}`"
      @click="handlePaginationUpdate(number)"
      @keydown.enter="handlePaginationUpdate(number)"
    >
      {{ number + 1 }}
    </button>
    <button
      id="button-to-next-page"
      class="pagination__button"
      :aria-label="`Click to select the next page`"
      :disabled="selectedPage === pageCount"
      @click="handlePaginationUpdate(selectedPage + 1)"
      @keydown.enter="handlePaginationUpdate(selectedPage + 1)"
    >
      &gt;
    </button>
    <button
      id="button-to-last-page"
      class="pagination__button"
      :aria-label="`Click to select last page`"
      :disabled="selectedPage === pageCount"
      @click="handlePaginationUpdate(pageCount)"
      @keydown.enter="handlePaginationUpdate(pageCount)"
    >
      &gt;&gt;
    </button>
  </div>
</template>

<script src="./Pagination.js"></script>

<style scoped lang="scss" src="./Pagination.scss"></style>
