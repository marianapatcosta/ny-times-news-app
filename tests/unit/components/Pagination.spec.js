import { mount } from "@vue/test-utils";
import "../match-media-mock";
import Pagination from "@/components/pagination/Pagination.vue";

describe("Pagination.vue", () => {
  const pageCount = 10;
  let wrapper;
  beforeEach(() => {
    wrapper = mount(Pagination, {
      props: { pageCount },
    });
  });

  afterEach(() => {
    wrapper.unmount();
  });
  it("should render without errors and match snapshot renders correctly", () => {
    expect(wrapper.element).toMatchSnapshot();
  });

  it("should emit 'update-pagination' event when a pagination bar button is clicked", async () => {
    const paginationButtons = wrapper.findAll("button");
    await paginationButtons.at(3).trigger("click");
    expect(wrapper.emitted()).toHaveProperty("update-pagination");
  });

  it("should emit 'update-pagination' event with pageCount value when 'button-to-last-page' is clicked", async () => {
    const button = wrapper.find("#button-to-last-page");
    await button.trigger("click");
    expect(wrapper.emitted()["update-pagination"][0]).toEqual([10]);
  });

  it("should emit 'update-pagination' event with pageCount value when 'button-to-last-page' is clicked", async () => {
    const button = wrapper.find("#button-to-last-page");
    await button.trigger("click");
    expect(wrapper.emitted()["update-pagination"][0]).toEqual([10]);
  });

  it("display 0, 1, 2, 3, 4 buttons when 'button-to-last-page' is clicked", async () => {
    expect(wrapper.vm.paginationNumbers).toEqual([0, 1, 2, 3, 4]);
  });

  it("display 1, 2, 3, 4, 5 buttons when 'button-to-next-page' is clicked", async () => {
    const button = wrapper.find("#button-to-next-page");
    await button.trigger("click");
    await button.trigger("click");
    await button.trigger("click");
    expect(wrapper.vm.paginationNumbers).toEqual([1, 2, 3, 4, 5]);
  });

  it("display 6, 7, 8, 9, 10 buttons when 'button-to-last-page' is clicked", async () => {
    const button = wrapper.find("#button-to-last-page");
    await button.trigger("click");
    expect(wrapper.vm.paginationNumbers).toEqual([6, 7, 8, 9, 10]);
  });
});
