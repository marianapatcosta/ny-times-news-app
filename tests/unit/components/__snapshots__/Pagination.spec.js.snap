// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pagination.vue should render without errors and match snapshot renders correctly 1`] = `
<div
  class="pagination"
  data-cy="pagination"
>
  <button
    aria-label="Click to select first page"
    class="pagination__button"
    disabled=""
    id="button-to-first-page"
  >
     &lt; &lt; 
  </button>
  <button
    aria-label="Click to select the previous page"
    class="pagination__button"
    disabled=""
    id="button-to-previous-page"
  >
     &lt; 
  </button>
  
  <button
    aria-label="Click to select page 1"
    class="pagination__button pagination__button--active"
  >
    1
  </button>
  <button
    aria-label="Click to select page 2"
    class="pagination__button"
  >
    2
  </button>
  <button
    aria-label="Click to select page 3"
    class="pagination__button"
  >
    3
  </button>
  <button
    aria-label="Click to select page 4"
    class="pagination__button"
  >
    4
  </button>
  <button
    aria-label="Click to select page 5"
    class="pagination__button"
  >
    5
  </button>
  
  <button
    aria-label="Click to select the next page"
    class="pagination__button"
    id="button-to-next-page"
  >
     &gt; 
  </button>
  <button
    aria-label="Click to select last page"
    class="pagination__button"
    id="button-to-last-page"
  >
     &gt;&gt; 
  </button>
</div>
`;
